### DISCUSSION TITLE
Disable Scroll on new router in nextjs (Nextjs 13)

### DISCUSSION THREAD

**Discussion by okaforcj** [vercel/next.js#47781] [Help]
### Summary

It seems like the previous api to prevent scroll has been removed for next.js 13 i.e. before we could do something lilke `router.push("path",undefined, {scroll: false}` but now the only option is 'forceOptimisticUpdate'. Note: the Link component still has the option to disable scroll but its not the right property for my use case. Does anyone know how i can disable scroll to top using the router as it was before

**Comment by 18601673727** [comment]
To devs: Not only scroll to top need to be disabled, but also scroll to left needs a way too, I'm having a overflow-x list, everytime user clicks link inside of it, the scroll x position reset to 0, tried both `<Link>` and `router.push()` both not working as expected.

**Reply by satu-jovanhartono** [reply]
is there any solution to this? am facing the same issue right now

**Comment by matheusquintaes** [comment]
When utilizing the [Intercepting Routes feature to create modals](https://beta.nextjs.org/docs/routing/intercepting-routes#modals), the page scrolls to the top, which is bad to usability.

For example, consider a masonry grid with photos, similar to Instagram's explore page or Pinterest. When a user clicks on a photo, a modal opens with shareable content via a URL, while the page's scroll position remains unchanged.

However, when using Next.js 13, as demonstrated in the [nextgram by the Vercel Team](https://nextgram.vercel.app/), which employs the new "Parallel Routes and Intercepting Routes" features, if a user scrolls to the last photo and clicks on it, the scroll position resets, negatively impacting the user experience.

**Comment by ettysekhon** [comment]
I am facing the same issue with the app router. I want to add a load more button to add more items on the page and also update the query parameters to reflect the page. It does not matter whether I use the `Link` component with scroll prop set to false or use router.push it scrolls up each time:

https://next13-cypress-music-search.vercel.app
https://github.com/ettysekhon/next13-cypress-music-search

**Reply by okanji** [reply]
This is a really big issue for me, I have a paginated table and it's really bad user experience to have the page scrolled to the top each time someone clicks the next or previous button. We need some visibility on this issue.

**Reply by JamesBliss** [reply]
@ettysekhon I've created a repo to implement a "fix" for this issue. Hopefully, it will help tide you over until a proper solution from Next is released.

[Repo](https://github.com/JamesBliss/next13-cypress-music-search)
[Example](https://next13-cypress-music-search-ebon.vercel.app)

Key changes are:

- Created a [hook](https://github.com/JamesBliss/next13-cypress-music-search/blob/main/src/hooks/useNavigationEvent.ts) to watch for URL changes and then have support for a callback function. This code is modified from what was put forward in this [thread](https://github.com/vercel/next.js/discussions/42016)

- Updated the [MusicCardList](https://github.com/JamesBliss/next13-cypress-music-search/blob/main/src/components/MusicCardList.tsx) to support storing a scroll offset when the load more link is clicked and then use that stored value after the navigation event has run.

**Reply by webbiffy** [reply]
@okanji same issue on my end. next version 13.4.12
Have you manage to fix the issue?

**Reply by okanji** [reply]
> @okanji same issue on my end. next version 13.4.12
> 
> Have you manage to fix the issue?

No don't have a solution unfortunately

**Reply by p00000001** [reply]
Apparently this was resolved in https://github.com/vercel/next.js/issues/50105#issuecomment-1619782873 - have you tried with the scroll options setting in the latest version?

**Reply by timneutkens** [reply]
`<Link scroll={false}>` as well as `router.push('/somewhere', {scroll: false})` are both supported in the latest version.

**Reply by webbiffy** [reply]
In my scenario, I have a custom pagination component which has custom Button component (Prev and Next) to navigate the page. I have a useEffect which has a dependency of route.query, inside that useEffect Im using route.replace.

Everything works fine except that when I click the prev or next button, it still scrolls to top even tho I added the { scroll: false } on route.replace.

**Reply by rassemdev** [reply]
Still not working, tried using navigating to same page with extra query params. it is scrolling back to top.

import { useRouter } from 'next/router';
const router = useRouter();
router.push(`/blogs?search=${searchParam.toString()}&limit=${limit}`, { scroll: false });

**Reply by hadnet** [reply]
I'm using 13.4.1 and it's not available. Is it available in what version?
PS: Ok, it's the >=13.4.9 version

**Reply by james-william-r** [reply]
FYI: Also works on `.replace()` ✨

**Reply by alessiomaffeis** [reply]
@timneutkens `router.push('/somewhere', {scroll: false}) ` still doesn't seem to restore scroll position (latest version, app router). By contrast using `router.back()` correctly restores the scroll position.

**Reply by alessiomaffeis** [reply]
Edit: the actual behaviour for `scroll: false` was clarified in: https://github.com/vercel/next.js/issues/57258

It seems that it merely keeps the outgoing route's scroll position, rather than restoring the previous scroll position on the incoming route.

Perhaps this could be clarified a bit in the docs: https://nextjs.org/docs/app/api-reference/functions/use-router#disabling-scroll-restoration

**Reply by alessiomaffeis** [reply]
I suggested a minor tweak to address that: https://github.com/vercel/next.js/pull/57273/files

**Comment by HenryBrown0** [comment]
Related issue: #50105

**Comment by cphoover** [comment]
Also seeing this issue...

**Reply by piedrafita8** [reply]
Same problem here, i dont have a solution yet

**Comment by makhdum530** [comment]
router.push(paths.setting.detaills(btoa(row.id), pagination.pageIndex + 1, pagination.pageSize),{scroll: false})
Not working  scroll false 
in table data

**Comment by martinbartolo** [comment]
Also getting this issue on App Router using 14.0.1.

I've tried using a Link tag with scroll={false} or a router.replace with scroll={false} and my pages scroll to the top anyway in both cases.

**Reply by alessiomaffeis** [reply]
That is actually the intended behaviour:

https://github.com/vercel/next.js/discussions/47781#discussioncomment-7361088

**Reply by channyeintun** [reply]
> Also getting this issue on App Router using 14.0.1.
> 
> I've tried using a Link tag with scroll={false} or a router.replace with scroll={false} and my pages scroll to the top anyway in both cases.

I made a workaround.
```javascript
import { useEffect, useLayoutEffect } from "react";
import { usePathname, useSearchParams } from "next/navigation";

export default function useScrollRestoration() {
    const pathname = usePathname();
    const searchParams = useSearchParams();

    const url = `${pathname}?${searchParams}`;

    useLayoutEffect(() => {
        if (typeof window !== "undefined") {
            const beforeUnloadHandler = () => sessionStorage.clear();

            window.addEventListener("beforeunload", beforeUnloadHandler);

            if ("history" in window && "scrollRestoration" in history) {
                history.scrollRestoration = "manual";
            }

            const scrollY = sessionStorage.getItem(url);

            window.scrollTo(0, +scrollY);

            return () => {
                window.removeEventListener("beforeunload", beforeUnloadHandler);
            };
        }
    }, [url]);

    useEffect(() => {
        let scrollTimeout;
        let lastScrollTop = window.scrollY;

        const scrollHandler = () => {
            clearTimeout(scrollTimeout);
            scrollTimeout = setTimeout(() => {
                if (window.scrollY !== lastScrollTop) {
                    lastScrollTop = window.scrollY;
                    sessionStorage.setItem(url, window.scrollY);
                }
            }, 0);
        };

        const scrollEndHandler = () => {
            sessionStorage.setItem(url, window.scrollY);
        };

        if ("onscrollend" in window) {
            // Browser supports scrollend
            document.addEventListener("scrollend", scrollEndHandler);
        } else {
            // Polyfill for browsers that don't support scrollend
            document.addEventListener("scroll", scrollHandler);
        }

        return () => {
            if ("onscrollend" in window) {
                document.removeEventListener("scrollend", scrollEndHandler);
            } else {
                document.removeEventListener("scroll", scrollHandler);
                clearTimeout(scrollTimeout);
            }
        };
    }, [url]);
}
```

**Reply by pablojsx** [reply]
do you get a page jump with this hook? an useeffect will jump and it looks even uglier
